### DISCUSSION TITLE
Releases tag `latest`

### DISCUSSION THREAD

**Discussion by sjackson0109** [Set-OutlookSignatures/Set-OutlookSignatures#92] [General]
@GruberMarkus  - upon your next product release, can you add the tag `latest` and then remove the tag from the previous release.

This means we can automate the download process using:
https://github.com/Set-OutlookSignatures/Set-OutlookSignatures/releases/download/latest/Set-OutlookSignatures.zip
instead of:
https://github.com/Set-OutlookSignatures/Set-OutlookSignatures/releases/download/v4.7.0/Set-OutlookSignatures_v4.7.0.zip

**Comment by sjackson0109** [comment]
Finding the tags:
```powershell
$repo1 = "Set-OutlookSignatures/Set-OutlookSignatures"
$file1 = "Set-OutlookSignatures.zip"
$version1 = "latest" #Options: "latest", "v1.0.0" etc
$releases1 = "https://api.github.com/repos/$repo1/releases"
(Invoke-WebRequest $releases1 | ConvertFrom-Json).tag_name
```
Result:
```output
v4.7.0
v4.6.1
v4.6.0
v4.5.0
v4.4.0
v4.3.0
v4.2.1
v4.2.0
v4.1.0
v4.0.0
```

**Reply by sjackson0109** [reply]
I could always strip the leading `v` character, and then sort the numbers.. or possibly always download array item `0`...
But the `latest` tag would be preferrable.

**Reply by sjackson0109** [reply]
I might have to use this for now:
```
(Invoke-WebRequest "https://api.github.com/repos/Set-OutlookSignatures/Set-OutlookSignatures/tags" | ConvertFrom-Json)[0]
```
returns
```
name        : v4.7.0
zipball_url : https://api.github.com/repos/Set-OutlookSignatures/Set-OutlookSignatures/zipball/refs/tags/v4.7.0
tarball_url : https://api.github.com/repos/Set-OutlookSignatures/Set-OutlookSignatures/tarball/refs/tags/v4.7.0
commit      : @{sha=bd63fc1b68cad52811ea64c3adf291cf11365037; url=https://api.github.com/repos/Set-OutlookSignatures/Set-OutlookSignatures/commits/bd63fc1b68cad52811ea64c3adf291cf11365037}
node_id     : MDM6UmVmMzU0NTkwMjk0OnJlZnMvdGFncy92NC43LjA=
```

**Reply by sjackson0109** [reply]
download link:

Invoke-WebRequest "$($(Invoke-WebRequest "https://api.github.com/repos/Set-OutlookSignatures/Set-OutlookSignatures/tags" | ConvertFrom-Json)[0].zipball_url)" -Out "$temp/Set-OutlookSignatures.zip"

**Comment by GruberMarkus** [comment]
Hi @sjackson0109,

I will not add an additional "latest" tag, because i already set a GitHub marker for the latest release.

https://github.com/Set-OutlookSignatures/Set-OutlookSignatures/releases/latest will always redirect you to the latest non-pre-release.

I would not recommend automatically downloading the latest release, but to define a specific version in your script. This way, you have a chance to read the changelog and to test a newer release before running it in a production environment. Also, the name of the download files may change, or there may be multiple files.

If you want to go this way, I would use something like the following PowerShell code to determine the download link for the latest non-pre-release:
```
((Invoke-WebRequest -Uri 'https://api.github.com/repos/Set-OutlookSignatures/Set-OutlookSignatures/releases/latest' -UseBasicParsing | ConvertFrom-Json).assets | Where-Object { $_.name -ilike 'Set-OutlookSignatures_v*.zip' })[0].browser_download_url
```
While writing this, this gives `https://github.com/Set-OutlookSignatures/Set-OutlookSignatures/releases/download/v4.7.0/Set-OutlookSignatures_v4.7.0.zip` as download location.



Regarding version sorting: Keep in mind that Set-OutlookSignatures uses semantic versioning. This means you can not just sort by numbers, as there may appear letters and release states in the tags, for example
```
1.0.0-alpha-a.b-c-somethinglong+build.1-aef.1-its-okay
1.0.0+0.build.1-rc.10000aaa-kk-0.1
1.0.0-rc.1+build.1
1.0.0
1.2.3-beta
1.2.3
2.0.0-rc.1+build.123
2.0.0
2.0.0+build.1848
2.0.1-alpha.1227
```

That's why I use the GitHub "latest" marker mentioned above - you can be sure that https://github.com/Set-OutlookSignatures/Set-OutlookSignatures/releases/latest will always redirect you to the latest non-pre-release.

**Comment by GruberMarkus** [comment]
I see that https://github.com/alltimeuk/EmailTemplates/blob/main/Intune/Remediate-EmailSignatures.ps1 still sticks to the not recommended method of parsing tags and using zipball_url, instead of using the code I recommended above.

Your current solution:
- Downloads source code, not a release. It works for the tag v4.7.0, but not for tags before or afterwards, because of a specific problem with a missing .gitattributes file in v4.7.0 (which has been corrected with a separate commit a few days after the release of v4.7.0). When downloading any release before or after v4.7.0, you will only see a file named "download information.txt" with a textual hint to download a release or to clone the repository. In other words: In every release but v4.7.0, the downloaded zipball_url will be nearly empty.
- Does not consider semantic versioning. It will blindly download the latest release, no matter if there are breaking changes or if the release is an alpha or beta version.


As recommended in my last post, the following one-liner downloads the release (not the zipball) marked as latest release:
```
((Invoke-WebRequest -Uri 'https://api.github.com/repos/Set-OutlookSignatures/Set-OutlookSignatures/releases/latest' -UseBasicParsing | ConvertFrom-Json).assets | Where-Object { $_.name -ilike 'Set-OutlookSignatures_v*.zip' })[0].browser_download_url
```

Personally, I would still prefer defining a specific release for download, as it allows me to test a new release before I deploy it to my users. You just need a simple change to the code above to accomplish this:
```
$githubProductOrg = "Set-OutlookSignatures"
$githubProductRepo = "Set-OutlookSignatures"
$githubProductRelease = 'v4.5.0'

(((Invoke-WebRequest -Uri "https://api.github.com/repos/$githubProductOrg/$githubProductRepo/releases" -UseBasicParsing | ConvertFrom-Json) | where {$_.tag_name -ieq $githubProductRelease }).assets | Where-Object { $_.name -ilike 'Set-OutlookSignatures_v*.zip'})[0].browser_download_url
```


Please provide feedback if this solves your request.

**Comment by sjackson0109** [comment]
Thanks, I’ll take that into consideration. Not had any time since our last messageSimonOn 9 Nov 2023, at 08:29, Markus Gruber ***@***.***> wrote:﻿
I see that https://github.com/alltimeuk/EmailTemplates/blob/main/Intune/Remediate-EmailSignatures.ps1 still sticks to the not recommended method of parsing tags and using zipball_url, instead of using the code I recommended above.
Your current solution:

Downloads source code, not a release. It works for the tag v4.7.0, but not for tags before or afterwards, because of a specific problem with a missing .gitattributes file in v4.7.0 (which has been corrected with a separate commit a few days after the release of v4.7.0). When downloading any release before or after v4.7.0, you will only see a file named "download information.txt" with a textual hint to download a release or to clone the repository. In other words: In every release but v4.7.0, the downloaded zipball_url will be nearly empty.
Does not consider semantic versioning. It will blindly download the latest release, no matter if there are breaking changes or if the release is an alpha or beta version.

As recommended in my last post, the following one-liner downloads the release (not the zipball) marked as latest release:
((Invoke-WebRequest -Uri 'https://api.github.com/repos/Set-OutlookSignatures/Set-OutlookSignatures/releases/latest' -UseBasicParsing | ConvertFrom-Json).assets | Where-Object { $_.name -ilike 'Set-OutlookSignatures_v*.zip' })[0].browser_download_url

Personally, I would still prefer defining a specific release for download, as it allows me to test a new release before I deploy it to my users. You just need a simple change to the code above to accomplish this:
$githubProductOrg = "Set-OutlookSignatures"
$githubProductRepo = "Set-OutlookSignatures"
$githubProductRelease = 'v4.5.0'

(((Invoke-WebRequest -Uri "https://api.github.com/repos/$githubProductOrg/$githubProductRepo/releases" -UseBasicParsing | ConvertFrom-Json) | where {$_.tag_name -ieq $githubProductRelease }).assets | Where-Object { $_.name -ilike 'Set-OutlookSignatures_v*.zip'})[0].browser_download_url

Please provide feedback if this solves your request.

**Reply by sjackson0109** [reply]
Hi again.  Not had any time lately to consider coming back to this until now..

Appreciate you spending time writing webrequests for me to poll to get the browser_download_url. Very useful.
Only issue i have is the `latest` release provided by github doesn't seem to have a SHA checksum parameter, so we can download and compare to what was published. I doubt this is going to introduce any real-risk; just a bug-bearer I have.

i'll be updating my code today.

**Comment by GruberMarkus** [comment]
Starting with v4.9.0, you can use .\docs\releases.txt to get the version of the already downloaded code, and, for example, compare this with the tag of the latest release.

The next release will contain a new FAQ with sample code showing how to use Set-OutlookSignatures with Intune, inspired by this discussion and your public code.

The setup is straight-forward:
1. Check script
   - Fail when a defined log file does not exist or is older than X hours
3. Remediation script
   - If not already installed, download and "install" (extract) a defined version of Set-OutlookSignatures. 
   - Run Set-OutlookSignatures with defined parameters and log the output for use in the Check script
   - Exit with code 1 if anything goes wrong during the whole process

It will very likely look as follows. @sjackson0109: I would love to hear your feedback.


## 17.31. How can I deploy and run Set-OutlookSignatures using Microsoft Intune? 
There are multiple ways to integrate Set-OutlookSignatures in Intune, depending on your configuration.

When not using an Always On VPN, place your configuration and template files in a SharePoint document library that can be accessed from the internet.
### 17.31.1. Application package
The classic way is to deploy an application package. You can use tools such as [IntuneWin32App](https://github.com/MSEndpointMgr/IntuneWin32App) for this.

As Set-OutlookSignatures does not have a classic installer, you will have to create a small wrapper script that simulates an installer. You will have to update the package or create a new one with every new release you plan to use - just as with any other application you want to deploy.

Deployment is only the first step, as the software needs to be run regularly. You have multiple options for this: Let the user run it via a start menu entry or a desktop shortcut, use scheduled tasks, a background service, or a remediation script (which is probably the most convenient way to do it).
### 17.31.2. Remediation script
With remediation, you have two scripts: One checking for a certain status, and another one running when the detection script exits with an error code of 1.

Remediation scripts can be configured to run in the context of the current user, and you can define how often they should run.

For Set-OutlookSignatures, you could use the following scripts, which do not require the creation and deployment of an application package as an additional benefit.

Detect-SetOutlookSignatures.ps1 could look like this (sample code):
```
# Log file for Set-OutlookSignatures, must be identical in detection and remediation script
$logFile = 'c:\path\to\the\user\specific\logfile.txt'

# Interval in hours between runs of Set-OutlookSignatures
$maximumAgeHours = 2


If (-not (Test-Path $logFile)) {
    Write-Error 'Log file not found, Set-OutlookSignatures has not yet run.'
    Write-Error 'Exit with error code 1 to trigger remediation script.'

    exit 1
} else {
    If ((Get-Date).AddHours(-$maximumAgeHours) -ge (Get-Item -LiteralPath $logFile).LastWriteTime) {
        Write-Error "Log file found, it is at least $maximumAgeHours hours old."
        Write-Error 'Exit with error code 1 to trigger remediation script.'

        exit 1
    } Else {
        Write-Host "Log file found, it is younger than $maximumAgeHours hours."
        Write-Host 'Exit with error code 0 to not trigger remediation script.'

        exit 0
    }
}
```

Remediate-SetOutlookSignatures.ps1 could look like this (sample code):
```
# Log file for Set-OutlookSignatures, must be identical in detection and remediation script
$logFile = 'c:\path\to\the\user\specific\logfile.txt'

# Version of Set-OutlookSignatures to use/download
$versionToUse = 'v4.10.0'

# Where to find Set-OutlookSignatures locally
$softwarePath = 'c:\path\to\Set-OutlookSignatures'

# Parameters for the later execution of Set-OutlookSignatures.ps1
$parameters = @{
    SignatureTemplatePath       = 'https://URI/to/SharePoint/Libary/with/Templates' # Path to SharePoint document library containing the templates, ini files, ...
    BenefactorCircleId          = 'xxx'
    BenefactorCircleLicenseFile = 'xxx'
}


try {
    # Get currently installed version
    $currentVersion = $null

    if (-not (Test-Path $softwarePath)) {
        New-Item -Path $softwarePath -ItemType Directory
    } else {
        try {
            $currentVersion = @(Get-Content -LiteralPath (Join-Path -Path $softwarePath -ChildPath 'docs\releases.txt'))[-1]
        } catch {
            $currentVersion = $null
        }
    }

    # Install Set-OutlookSignatures, if not already available in the required version
    if ($currentVersion -ine $versionToUse) {
        $tempFile = New-TemporaryFile | Rename-Item -NewName { [IO.Path]::ChangeExtension($_, '.zip') } -PassThru

        Invoke-WebRequest -Uri "https://github.com/Set-OutlookSignatures/Set-OutlookSignatures/releases/download/$($versionToUse)/Set-OutlookSignatures_$($versionToUse).zip" -UseBasicParsing -OutFile $tempFile

        Add-Type -Assembly System.IO.Compression.FileSystem

        $zip = [IO.Compression.ZipFile]::OpenRead($tempFile)

        $entries = $zip.Entries | Where-Object { $_.FullName -ilike "Set-OutlookSignatures_$($versionToUse)/*" } | Sort-Object

        $entries | ForEach-Object {
            $dest = $(Join-Path -Path $softwarePath -ChildPath ($_.FullName -ireplace "^$([regex]::escape("Set-OutlookSignatures_$($versionToUse)/"))"))

            if (($_.FullName.EndsWith('/')) -or (-not (Test-Path (Split-Path $dest)))) {
                $null = New-Item -Path $dest -ItemType Directory -Force
            } else {
                [IO.Compression.ZipFileExtensions]::ExtractToFile($_, $dest, $true)
            }
        }

        $zip.Dispose()

        Remove-Item -LiteralPath $tempFile -Force
    }


    # Run Set-OutlookSignatures
    Start-Transcript -LiteralPath $logFile -Force # Required for detection script

    & (Join-Path -Path $softwarePath -ChildPath 'Set-OutlookSignatures.ps1') @parameters

    Stop-Transcript
} catch {
    Write-Error $error[0]

    exit 1
}
```
